### PR TITLE
register float64 GPU kernel for Conv3d

### LINKED ISSUE #21295: support 3d convolution with float64
------------------------

### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:
No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
Windows 7
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**:
N/A
- **TensorFlow installed from (source or binary)**:
from pip install

- **TensorFlow version (use command below)**:
tensorflow 1.9.0

- **Python version**:
3.6
- **Bazel version (if compiling from source)**:
- **GCC/Compiler version (if compiling from source)**:
- **CUDA/cuDNN version**:
cuDNN 9.0 v7, CUDA9.1

- **GPU model and memory**:
GTX980Ti 6GB, Geforce Titan X 12GB

- **Exact command to reproduce**:

You can collect some of this information using our environment capture script:

https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh

You can obtain the TensorFlow version with

python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"

### Describe the problem
Describe the problem clearly here. Be sure to convey here why it's a bug in TensorFlow or a feature request.

I'm working on some convolutional calculation on the Tensorflow and got some strange errors with the Double Precision numbers.
I needed to calculate those tensors in float64 because of the precision issues.

The conv3d works fine with me on float32, but on the Double precision, it only works on CPU and on the GPU it causes error.

So is there any way I can get access to this operation?

### Source code / logs
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached. Try to provide a reproducible test case that is the bare minimum necessary to generate the problem.

import tensorflow as tf
import numpy as np

with tf.device('/device:GPU:0'):

    a = tf.constant(2 * np.ones([1, 200, 200, 200, 9], dtype=np.float64), dtype=tf.float64)

    b = tf.constant(np.ones([3, 3, 3, 9, 9], dtype=np.float64), dtype=tf.float64)

    sess = tf.Session()

    init = tf.global_variables_initializer()

    sess.run(init)

    c = tf.nn.conv3d(a, b, strides=[1, 1, 1, 1, 1], padding='SAME')
    print(a)
    print(b)
    print(c)
    for i in range(1000):
        print(sess.run(c))

the output says

InvalidArgumentError (see above for traceback): Cannot assign a device for operation 'Conv3D': Could not satisfy explicit device specification '/device:GPU:0' because no supported kernel for GPU devices is available.
Registered kernels:
  device='CPU'; T in [DT_HALF]
  device='CPU'; T in [DT_FLOAT]
  device='CPU'; T in [DT_DOUBLE]
  device='GPU'; T in [DT_HALF]
  device='GPU'; T in [DT_FLOAT]

 [[Node: Conv3D = Conv3D[T=DT_DOUBLE, data_format="NDHWC", dilations=[1, 1, 1, 1, 1], padding="SAME", strides=[1, 1, 1, 1, 1], _device="/device:GPU:0"](Const, Const_1)]]

### PR DESCRIPTION
**Pull request by facaiy**
Fix #21295.

### CODE REVIEW

**Comment by case540** [comment]
Pinging review

**Review by josh11b — DISMISSED** [review]

**Comment by facaiy** [comment]
Any update?

**Comment by facaiy** [comment]
@yifeif Hi, yifei. Do you have any update for the PR? Thank you.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @case540: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by facaiy** [comment]
@josh11b Hi, what can I do for the PR?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @case540: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by facaiy** [comment]
ping @josh11b 

@case540 @yifeif could you give a hand? Thanks.

**Comment by yifeif** [comment]
Hi Josh, the change and test results from previous import are at cl/208243530. We might need to do anther import.

**Comment by facaiy** [comment]
Thank you, @yifeif @josh11b . Is it necessary to merge / rebase the latest master branch?

**Comment by drpngx** [comment]
This appears to be failing the test. Triggering again.

**Comment by facaiy** [comment]
Thanks, I'll check the log of failed test later.

**Comment by drpngx** [comment]
```
2018-10-15 18:28:30.260444: E tensorflow/core/common_runtime/executor.cc:624] Executor failed to create kernel. Invalid argument: Conv3DBackpropInputOpV2 only supports NDHWC on the CPU.
```
seems to be triggering.

**Comment by facaiy** [comment]
@drpngx Hi, I registered double kernel for gradient op. Unfortunately, build failed. Do you have any suggestion to fix it ?

```bash
ERROR: /apsarapangu/disk1/facai.yfc/workshop/tensorflow/gpu/tensorflow/core/kernels/BUILD:3463:1: C++ compilation of rule '//tensorflow/core/kernels:conv_ops' failed (Exit 1)
In file included from external/eigen_archive/unsupported/Eigen/CXX11/Tensor:112:0,
                 from ./third_party/eigen3/unsupported/Eigen/CXX11/Tensor:1,
                 from ./tensorflow/core/framework/tensor_types.h:19,
                 from ./tensorflow/core/kernels/conv_3d.h:21,
                 from tensorflow/core/kernels/conv_grad_ops_3d.cc:19:
external/eigen_archive/unsupported/Eigen/CXX11/src/Tensor/TensorEvaluator.h: In instantiation of 'bool Eigen::TensorEvaluator<Derived, Device>::evalSubExprsIfNeeded(Eigen::TensorEvaluator<Derived, Device>::CoeffReturnType*) [with Derived = Eigen::TensorMap<Eigen::Tensor<double, 5, 1, int>, 16, Eigen::MakePointer>; Device = Eigen::GpuDevice; Eigen::TensorEvaluator<Derived, Device>::CoeffReturnType = double]':
external/eigen_archive/unsupported/Eigen/CXX11/src/Tensor/TensorAssign.h:124:5:   required from 'bool Eigen::TensorEvaluator<const Eigen::TensorAssignOp<LhsXprType, RhsXprType>, Device>::evalSubExprsIfNeeded(Eigen::TensorEvaluator<const Eigen::TensorAssignOp<LhsXprType, RhsXprType>, Device>::Scalar*) [with LeftArgType = Eigen::TensorMap<Eigen::Tensor<double, 5, 1, int>, 16, Eigen::MakePointer>; RightArgType = const Eigen::TensorPaddingOp<const Eigen::array<Eigen::IndexPair<int>, 5ul>, const Eigen::TensorMap<Eigen::Tensor<const double, 5, 1, int>, 16, Eigen::MakePointer> >; Device = Eigen::GpuDevice; Eigen::TensorEvaluator<const Eigen::TensorAssignOp<LhsXprType, RhsXprType>, Device>::Scalar = double]'
external/eigen_archive/unsupported/Eigen/CXX11/src/Tensor/TensorExecutor.h:35:66:   required from 'static void Eigen::internal::TensorExecutor<Expression, Device, Vectorizable>::run(const Expression&, const Device&) [with Expression = const Eigen::TensorAssignOp<Eigen::TensorMap<Eigen::Tensor<double, 5, 1, int>, 16, Eigen::MakePointer>, const Eigen::TensorPaddingOp<const Eigen::array<Eigen::IndexPair<int>, 5ul>, const Eigen::TensorMap<Eigen::Tensor<const double, 5, 1, int>, 16, Eigen::MakePointer> > >; Device = Eigen::GpuDevice; bool Vectorizable = true]'
external/eigen_archive/unsupported/Eigen/CXX11/src/Tensor/TensorDevice.h:35:79:   required from 'Eigen::TensorDevice<ExpressionType, DeviceType>& Eigen::TensorDevice<ExpressionType, DeviceType>::operator=(const OtherDerived&) [with OtherDerived = Eigen::TensorPaddingOp<const Eigen::array<Eigen::IndexPair<int>, 5ul>, const Eigen::TensorMap<Eigen::Tensor<const double, 5, 1, int>, 16, Eigen::MakePointer> >; ExpressionType = Eigen::TensorMap<Eigen::Tensor<double, 5, 1, int>, 16, Eigen::MakePointer>; DeviceType = Eigen::GpuDevice]'
./tensorflow/core/kernels/conv_2d.h:254:19:   required from 'void tensorflow::functor::PadInput<Device, T, IndexType, NDIMS>::operator()(const Device&, typename tensorflow::TTypes<T, NDIMS, IndexType>::ConstTensor, const std::array<int, (NDIMS - 2)>&, const std::array<int, (NDIMS - 2)>&, typename tensorflow::TTypes<T, NDIMS, IndexType>::Tensor, tensorflow::TensorFormat) [with Device = Eigen::GpuDevice; T = double; IndexType = int; int NDIMS = 5; typename tensorflow::TTypes<T, NDIMS, IndexType>::ConstTensor = Eigen::TensorMap<Eigen::Tensor<const double, 5, 1, int>, 16, Eigen::MakePointer>; typename tensorflow::TTypes<T, NDIMS, IndexType>::Tensor = Eigen::TensorMap<Eigen::Tensor<double, 5, 1, int>, 16, Eigen::MakePointer>]'
tensorflow/core/kernels/conv_grad_ops_3d.cc:1638:65:   required from 'void tensorflow::Conv3DBackpropFilterOp<Eigen::GpuDevice, T>::Compute(tensorflow::OpKernelContext*) [with T = double]'
tensorflow/core/kernels/conv_grad_ops_3d.cc:1867:1:   required from here
external/eigen_archive/unsupported/Eigen/CXX11/src/Tensor/TensorEvaluator.h:59:7: error: invalid use of incomplete type 'const struct Eigen::GpuDevice'
       m_device.memcpy((void*)dest, m_data, sizeof(Scalar) * m_dims.TotalSize());
       ^
In file included from external/eigen_archive/unsupported/Eigen/CXX11/Tensor:92:0,
                 from ./third_party/eigen3/unsupported/Eigen/CXX11/Tensor:1,
                 from ./tensorflow/core/framework/tensor_types.h:19,
                 from ./tensorflow/core/kernels/conv_3d.h:21,
                 from tensorflow/core/kernels/conv_grad_ops_3d.cc:19:
external/eigen_archive/unsupported/Eigen/CXX11/src/Tensor/TensorForwardDeclarations.h:102:8: error: forward declaration of 'const struct Eigen::GpuDevice'
 struct GpuDevice;
        ^
In file included from external/eigen_archive/unsupported/Eigen/CXX11/Tensor:112:0,
                 from ./third_party/eigen3/unsupported/Eigen/CXX11/Tensor:1,
                 from ./tensorflow/core/framework/tensor_types.h:19,
                 from ./tensorflow/core/kernels/conv_3d.h:21,
                 from tensorflow/core/kernels/conv_grad_ops_3d.cc:19:
external/eigen_archive/unsupported/Eigen/CXX11/src/Tensor/TensorEvaluator.h: In instantiation of 'bool Eigen::TensorEvaluator<const Derived, Device>::evalSubExprsIfNeeded(Eigen::TensorEvaluator<const Derived, Device>::CoeffReturnType*) [with Derived = Eigen::TensorMap<Eigen::Tensor<const double, 5, 1, int>, 16, Eigen::MakePointer>; Device = Eigen::GpuDevice; Eigen::TensorEvaluator<const Derived, Device>::CoeffReturnType = const double]':
external/eigen_archive/unsupported/Eigen/CXX11/src/Tensor/TensorPadding.h:140:5:   required from 'bool Eigen::TensorEvaluator<const Eigen::TensorPaddingOp<PaddingDimensions, XprType>, Device>::evalSubExprsIfNeeded(Eigen::TensorEvaluator<const Eigen::TensorPaddingOp<PaddingDimensions, XprType>, Device>::Scalar*) [with PaddingDimensions = const Eigen::array<Eigen::IndexPair<int>, 5ul>; ArgType = const Eigen::TensorMap<Eigen::Tensor<const double, 5, 1, int>, 16, Eigen::MakePointer>; Device = Eigen::GpuDevice; Eigen::TensorEvaluator<const Eigen::TensorPaddingOp<PaddingDimensions, XprType>, Device>::Scalar = const double]'
external/eigen_archive/unsupported/Eigen/CXX11/src/Tensor/TensorAssign.h:129:62:   required from 'bool Eigen::TensorEvaluator<const Eigen::TensorAssignOp<LhsXprType, RhsXprType>, Device>::evalSubExprsIfNeeded(Eigen::TensorEvaluator<const Eigen::TensorAssignOp<LhsXprType, RhsXprType>, Device>::Scalar*) [with LeftArgType = Eigen::TensorMap<Eigen::Tensor<double, 5, 1, int>, 16, Eigen::MakePointer>; RightArgType = const Eigen::TensorPaddingOp<const Eigen::array<Eigen::IndexPair<int>, 5ul>, const Eigen::TensorMap<Eigen::Tensor<const double, 5, 1, int>, 16, Eigen::MakePointer> >; Device = Eigen::GpuDevice; Eigen::TensorEvaluator<const Eigen::TensorAssignOp<LhsXprType, RhsXprType>, Device>::Scalar = double]'
external/eigen_archive/unsupported/Eigen/CXX11/src/Tensor/TensorExecutor.h:35:66:   required from 'static void Eigen::internal::TensorExecutor<Expression, Device, Vectorizable>::run(const Expression&, const Device&) [with Expression = const Eigen::TensorAssignOp<Eigen::TensorMap<Eigen::Tensor<double, 5, 1, int>, 16, Eigen::MakePointer>, const Eigen::TensorPaddingOp<const Eigen::array<Eigen::IndexPair<int>, 5ul>, const Eigen::TensorMap<Eigen::Tensor<const double, 5, 1, int>, 16, Eigen::MakePointer> > >; Device = Eigen::GpuDevice; bool Vectorizable = true]'
external/eigen_archive/unsupported/Eigen/CXX11/src/Tensor/TensorDevice.h:35:79:   required from 'Eigen::TensorDevice<ExpressionType, DeviceType>& Eigen::TensorDevice<ExpressionType, DeviceType>::operator=(const OtherDerived&) [with OtherDerived = Eigen::TensorPaddingOp<const Eigen::array<Eigen::IndexPair<int>, 5ul>, const Eigen::TensorMap<Eigen::Tensor<const double, 5, 1, int>, 16, Eigen::MakePointer> >; ExpressionType = Eigen::TensorMap<Eigen::Tensor<double, 5, 1, int>, 16, Eigen::MakePointer>; DeviceType = Eigen::GpuDevice]'
./tensorflow/core/kernels/conv_2d.h:254:19:   required from 'void tensorflow::functor::PadInput<Device, T, IndexType, NDIMS>::operator()(const Device&, typename tensorflow::TTypes<T, NDIMS, IndexType>::ConstTensor, const std::array<int, (NDIMS - 2)>&, const std::array<int, (NDIMS - 2)>&, typename tensorflow::TTypes<T, NDIMS, IndexType>::Tensor, tensorflow::TensorFormat) [with Device = Eigen::GpuDevice; T = double; IndexType = int; int NDIMS = 5; typename tensorflow::TTypes<T, NDIMS, IndexType>::ConstTensor = Eigen::TensorMap<Eigen::Tensor<const double, 5, 1, int>, 16, Eigen::MakePointer>; typename tensorflow::TTypes<T, NDIMS, IndexType>::Tensor = Eigen::TensorMap<Eigen::Tensor<double, 5, 1, int>, 16, Eigen::MakePointer>]'
tensorflow/core/kernels/conv_grad_ops_3d.cc:1638:65:   required from 'void tensorflow::Conv3DBackpropFilterOp<Eigen::GpuDevice, T>::Compute(tensorflow::OpKernelContext*) [with T = double]'
tensorflow/core/kernels/conv_grad_ops_3d.cc:1867:1:   required from here
external/eigen_archive/unsupported/Eigen/CXX11/src/Tensor/TensorEvaluator.h:186:7: error: invalid use of incomplete type 'const struct Eigen::GpuDevice'
       m_device.memcpy((void*)data, m_data, m_dims.TotalSize() * sizeof(Scalar));
       ^
In file included from external/eigen_archive/unsupported/Eigen/CXX11/Tensor:92:0,
                 from ./third_party/eigen3/unsupported/Eigen/CXX11/Tensor:1,
                 from ./tensorflow/core/framework/tensor_types.h:19,
                 from ./tensorflow/core/kernels/conv_3d.h:21,
                 from tensorflow/core/kernels/conv_grad_ops_3d.cc:19:
external/eigen_archive/unsupported/Eigen/CXX11/src/Tensor/TensorForwardDeclarations.h:102:8: error: forward declaration of 'const struct Eigen::GpuDevice'
 struct GpuDevice;
        ^
In file included from ./tensorflow/core/kernels/conv_2d.h:23:0,
                 from tensorflow/core/kernels/conv_grad_ops_3d.cc:27:
./tensorflow/core/util/tensor_format.h: In function 'int tensorflow::GetTensorDimsFromSpatialDims(int, tensorflow::TensorFormat)':
./tensorflow/core/util/tensor_format.h:148:1: warning: control reaches end of non-void function [-Wreturn-type]
 }
 ^
./tensorflow/core/util/tensor_format.h: In function 'int tensorflow::GetTensorSpatialDims(int, tensorflow::TensorFormat)':
./tensorflow/core/util/tensor_format.h:124:1: warning: control reaches end of non-void function [-Wreturn-type]
 }
 ^
Target //tensorflow/python/kernel_tests:conv_ops_3d_test failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 48.441s, Critical Path: 47.49s
INFO: 273 processes: 273 local.
FAILED: Build did NOT complete successfully

FAILED: Build did NOT complete successfully
```

**Comment by facaiy** [comment]
@drpngx @josh11b Sorry for the delay. The environment was broken in my working station last month. I have fixed the failure on GPU Python3. And the failed test on GPU CC seems unrelated for me.

Could you take a look again? Thanks.

**Comment by facaiy** [comment]
@drpngx Hi, I merged the latest master branch to resolve code conflict. Those failures seems unrelated?

**Comment by drpngx** [comment]
OK, let's try to test internally.

**Comment by facaiy** [comment]
Thank you. Please keep me updated :-)
